test/contrib/test_socks.py

```diff
@@ -87,6 +90,24 @@ def _address_from_socket(sock):
         raise RuntimeError("Unexpected addr type: %r" % addr_type)
 
 
+def _set_up_fake_getaddrinfo(monkeypatch):
+    # Work around https://github.com/urllib3/urllib3/pull/2034
+    # Nothing prevents localhost to point to two different IPs. For example, in the
+    # Ubuntu set up by GitHub Actions, localhost points both to 127.0.0.1 and ::1.
+    #
+    # In case of failure, PySocks will try the same request on both IPs, but our
+    # handle_socks[45]_negotiation functions don't handle retries, which leads either to
+    # a deadlock or a timeout in case of a failure on the first address.
+    #
+    # However, some tests need to exercise failure. We don't want retries there, but
+    # can't affect PySocks retries via its API. Instead, we monkeypatch PySocks so that
+    # it only sees a single address, which effectively disables retries.
+    def fake_getaddrinfo(addr, port, family, socket_type):
+        return [(socket.AF_INET, socket_type, family, "", (addr, port))]
```

Comment:
Instead of hard-coding this value, can we instead filter out all non-`socket.AF_INET` to ensure there aren't any IPv6 connect attempts?

Reply:
Right, it would be an issue if localhost resolved to an IPv6 since we specify `AF_INET`. So I called `getaddrinfo()` and filtered out all non `socket.AF_INET`, and still returned a one-item list, which works locally.

However, I don't understand why I need the filtering! But I do. `::1` is first on my system, and TestSocks5Proxy.test_socks_with_invalid_password fails with `Failed to establish a new connection: [Errno 61] Connection refused'` if I don't filter out the IPv6. (The two other tests work fine.)

Reply:
Might be worth running Wireshark on your system to see which addresses, ports, etc get used when you run the test?